core/include/detray/core/detail/single_store copy.hpp

```diff
@@ -0,0 +1,151 @@
+    // TO DISCUSS: change naming to builder_type?
+    using host_type = vecmem::vector<T>;
+    using buffer_type = vecmem::data::vector_buffer<T>;
+    using view_type = vecmem::data::vector_view<T>;
+    using device_type = vecmem::device_vector<T>;
+
+    using size_type = typename device_type::size_type;
+    using link_type = size_type;
+    using single_link = link_type;
+    using range_link = std::array<link_type, 2>;
+};
+
+template <typename T, typename context_t>
+DETRAY_HOST
+class single_store_host : single_store_base<T, context_t> {
+
+    public:
+    using _ = single_store_base<T, context_t>;
+    
+    single_store_host(const typename _::size_type size) : _data(size) {};
+
+    // TO DISCUSS: Do we gain much, if anything from having a resource constructor here?
+
+    single_store_host(const typename _::size_type size, const T &arg) : _data(size, arg) {};
+
+    T& at(const typename _::link_type l) {
+        return _data.at(l);
+    }
+
+    typename _::size_type size() {
+        return static_cast<typename _::size_type>(_data.size());
+    }
+
+    typename _::view_type get_data() {
+        return vecmem::get_data(_data);
+    }
+
+    void reserve(const typename _::size_type size) {
+        _data.reserve(size);
+    }
+
+    void push_back(const T& value) {
+        _data.push_back(value);
+    }
+
+    private:
+    typename _::host_type _data;
+};
+
+template <typename T, typename context_t>
+DETRAY_HOST
+class single_store_buffer : single_store_base<T, context_t> {
+
+    public:
+    using _ = single_store_base<T, context_t>;
+    
+    single_store_buffer(typename _::size_type size, vecmem::memory_resource& resource) 
+        : _data(size, resource) {}
+
+    single_store_buffer(vecmem::memory_resource& resource)
+        : _data(resource) {}
+
+    single_store_buffer(typename _::view_type data, vecmem::memory_resource& resource, vecmem::copy& cpy)
+        : _data(data.size(), resource) {
+            cpy(data, _data)->wait();
+        }
+
+    typename _::view_type get_data() {
+        return vecmem::get_data(_data);
+    }
+
+    private:
+    typename _::buffer_type _data;
+};
+
+template <typename T, typename context_t>
+DETRAY_HOST_DEVICE
+class single_store_view : single_store_base<T, context_t> {
+
+    public:
+    using _ = single_store_base<T, context_t>;
+    
+    DETRAY_HOST_DEVICE
+    single_store_view(typename _::view_type data)
+        : _data(data) {}
+
+    DETRAY_HOST_DEVICE
+    T& at(typename _::link_type l) {
+        typename _::device_type dev(_data);
+        return dev.at(l);
+    }
+
+    DETRAY_HOST_DEVICE
+    typename _::size_type size() const {
+        return _data.size();
+    }
+
+    private:
+    typename _::view_type _data;
+};
+
+template<typename T, typename context_t = empty_context>
+struct single_store_types {
+    using host = single_store_host<T, context_t>;
+    using buffer = single_store_buffer<T, context_t>;
+    using view = single_store_view<T, context_t>;
+};
```

Comment:
So... We very well may want to go for such a UI in detray as well. While not perfect, I think this (https://github.com/acts-project/traccc/blob/main/core/include/traccc/edm/container.hpp#L127-L183) serves us reasonably well in traccc.

Implementation-wise I would debate things a bit though. I don't really see what the common base class brings to the table in this current version... :confused:

A common base class should be attempted for at least some of the classes to make code repetition a bit smaller. But I think the current common base doesn't do a whole lot for that. :frowning:

Reply:
>  I don't really see what the common base class brings to the table in this current version...

Single store is just a `vecmem::vector` (or single element tuple vector container) so yeah it is hard to do sth smarter than collection_types in traccc...